File: scripts/gen_lookup.py
#!/usr/bin/env python

# Imports
import sys
import csv
import socket
import SubnetTree

# Create a tree
tree = SubnetTree.SubnetTree()

# Create a new socket
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Connect to Team Cymru (https://www.team-cymru.org/IP-ASN-mapping.html)
s.connect(("whois.cymru.com", 43))

# Send the header
s.send("begin\nverbose\n")

# Make s into a file for ease of use
f = s.makefile()

# Read in the header and ignore it
f.readline()

# Use a DictReader to process the "file"
r = csv.DictReader(f,
	delimiter='|',
	quoting=csv.QUOTE_NONE,
	skipinitialspace=True,
	fieldnames=["AS", "Ignore", "ASPrefix", "ASCountry", "ASRegistry", "ASAllocDate", "ASName"]
)

# Setup a csv writer to stdout
w = csv.DictWriter(sys.stdout, fieldnames=["ASPrefix", "AS", "ASCountry", "ASRegistry", "ASAllocDate", "ASName"])
w.writeheader()

# For each IP
for line in sys.stdin:

	# Check if not already looked up in tree
	if line.strip() not in tree:

		# Look it up
		s.send(line.strip() + "\n")

		# Get the response from the socket
		result = next(r)

		# Delete
		del result["Ignore"]

		# Strip all the values since they come with whitespace
		result = { k:v.strip() for k, v in result.iteritems() }

		# Catch and ignore bad prefixs
		try: 		

			# Save the prefix in tree so we don't look it up twice for other IPs in the range
			tree[result["ASPrefix"]] = True

			# Print the prefix to stdout
			w.writerow(result)

		# Ignore the error
		except:
			pass

File: scripts/gen_lookup.go
package main

// Imports
import (
	"os"
	"net"
	"bufio"
	"strings"
	"strconv"
	"encoding/csv"
)

type CachedValue struct {
	Result []string
	Count int
}

// Hold all known cidr ranges
var cache = make(map[*net.IPNet]*CachedValue)

// Checks if a given ip is in cache
func lookupIP(ip net.IP) *CachedValue {

	// Loop each net, if it has ip return true
	for subnet, val := range cache {
		if subnet.Contains(ip) {
			return val
		}
	}

	// Else we don't know it
	return nil

}

// Entry point
func main() {

	// Threshold of ips before saving
	threshold, err := strconv.Atoi(os.Args[1])
	if err != nil {
		panic(err)
	}

	// Create a scanner so we can read in line by line
	scanner := bufio.NewScanner(bufio.NewReader(os.Stdin))

	// Resolve the address
	tcpAddr, err := net.ResolveTCPAddr("tcp", "whois.cymru.com:43")
	if err != nil {
		panic(err)
	}

	// Connect to the socket
	conn, err := net.DialTCP("tcp", nil, tcpAddr)
	if err != nil {
		panic(err)
	}

	// Send the begin and verbose
	_, err = conn.Write([]byte("begin\nverbose\n"))
	if err != nil {
		panic(err)
	}

	// Create a reader from the connection and read one line
	_, _, err = bufio.NewReader(conn).ReadLine()
	if err != nil {
		panic(err)
	}

	// Create a new csv reader
	csvReader := csv.NewReader(conn)
	csvReader.Comma = '|'

	// Create a csv writer and write the headers
	csvWriter := csv.NewWriter(os.Stdout)
	csvWriter.Write([]string{"AS", "ASPrefix", "ASCountry", "ASRegistry", "ASAllocDate", "ASName"})
	csvWriter.Flush()

	// Loop each line in file
	for scanner.Scan() {

		// Parse the line as an IP
		ip := net.ParseIP(scanner.Text())

		// Lookup the ip
		val := lookupIP(ip)

		// If it's not known
		if val == nil {

			// Request a lookup on that IP
			_, err = conn.Write([]byte(ip.String() + "\n"))
			if err != nil {
				panic(err)
			}

			// Read the result
			result, err := csvReader.Read()
			if err != nil {
				panic(err)
			}

			// Trim all values
			for key, val := range result {
				result[key] = strings.TrimSpace(val)
			}

			// Parse the response CIDR range
			_, cidr, err := net.ParseCIDR(result[2])
			if err != nil {
				panic(err)
			}

			// Make the ip list
			cache[cidr] = &CachedValue{
				Result: append(result[0:1], result[2:]...),
				Count: 0,
			}
			val = cache[cidr]

		}

		// If we passed the threshold
		if val.Count == threshold {

			// Write the new line
			csvWriter.Write(val.Result)
			csvWriter.Flush()

			// Increment the count to push over the threshold
			val.Count++

		} else {

			// Increment the count
			val.Count++

		}

	}
	
}